### PR TITLE
Audio: Improvement of multiband_drc memory allocation error handling.

### PR DESCRIPTION
**Pull request by ShriramShastry**
This check-in improves error handling in the `multiband_drc` component by enabling error logging and returning `-ENOMEM` if memory allocation fails. It replaces asserts with conditional checks to detect and log errors during memory copy operations. Finally, it streamlines the return logic in` multiband_drc_setup` and propagates the `init_coef` error code.

### CODE REVIEW

File: src/audio/multiband_drc/multiband_drc.c [hunk 1]
```diff
@@ -64,20 +64,25 @@ static void multiband_drc_reset_state(struct multiband_drc_state *state)
...
+		comp_err(dev, "Failed to allocate EQ coefficients");
 		return -ENOMEM;
+	}
 
 	/* Coefficients of the first biquad and second biquad */
 	ret = memcpy_s(eq->coef, sizeof(struct sof_eq_iir_biquad) * SOF_EMP_DEEMP_BIQUADS,
 		       coef, sizeof(struct sof_eq_iir_biquad) * SOF_EMP_DEEMP_BIQUADS);
-	assert(!ret);
+	if (ret) {
+		comp_err(dev, "Memory copy failed");
```
Comment (cujomalainey) [inline src/audio/multiband_drc/multiband_drc.c:83]: is there actually a high risk of this?

File: src/audio/multiband_drc/multiband_drc.c [hunk 2]
```diff
@@ -64,20 +64,25 @@ static void multiband_drc_reset_state(struct multiband_drc_state *state)
...
+static int multiband_drc_eq_init_coef_ch(struct comp_dev *dev, struct sof_eq_iir_biquad *coef,
 					 struct iir_state_df2t *eq)
 {
 	int ret;
 
 	eq->coef = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
 			   sizeof(struct sof_eq_iir_biquad) * SOF_EMP_DEEMP_BIQUADS);
-	if (!eq->coef)
+	if (!eq->coef) {
+		comp_err(dev, "Failed to allocate EQ coefficients");
```
Comment (cujomalainey) [inline src/audio/multiband_drc/multiband_drc.c:75]: the return code is logged by the caller which will point to this line, this is not needed

File: src/audio/multiband_drc/multiband_drc.c [hunk 3]
```diff
@@ -64,20 +64,25 @@ static void multiband_drc_reset_state(struct multiband_drc_state *state)
...
 		return -ENOMEM;
+	}
 
 	/* Coefficients of the first biquad and second biquad */
 	ret = memcpy_s(eq->coef, sizeof(struct sof_eq_iir_biquad) * SOF_EMP_DEEMP_BIQUADS,
 		       coef, sizeof(struct sof_eq_iir_biquad) * SOF_EMP_DEEMP_BIQUADS);
-	assert(!ret);
```
Comment (cujomalainey) [inline src/audio/multiband_drc/multiband_drc.c:80]: the assert is intentional, if memcpy broke we screwed up bad and should stop.

**Review by cujomalainey — COMMENTED** [review]
I don't see much value to this change currently, will need more convincing before I will sign off

File: src/audio/multiband_drc/multiband_drc.c [hunk 4]
```diff
@@ -205,10 +210,8 @@ static int multiband_drc_setup(struct processing_module *mod, int16_t channels,
 
 	/* Setup Crossover, Emphasis EQ, Deemphasis EQ, and DRC */
 	ret = multiband_drc_init_coef(mod, channels, rate);
-	if (ret < 0)
-		return ret;
-
-	return 0;
+	/* This will return 0 if no error occurred, or the negative error code. */
+	return ret;
```
Comment (lyakh) [inline src/audio/multiband_drc/multiband_drc.c:214]: if you're doing this, you could as well just make it
```
multiband_drc_init_coef(mod, channels, rate);
```
and the comment about the error code isn't needed - we have many functions in SOF, returning error codes, we don't need to explain this in each of them.

**Review by lgirdwood — COMMENTED** [review]
@ShriramShastry any update ? It looks like this PR can be simplified based on the review comments. Thanks

**Comment by ShriramShastry** [comment]
> @ShriramShastry any update ? It looks like this PR can be simplified based on the review comments. Thanks

yes, I'II update it shortly.

File: src/audio/multiband_drc/multiband_drc.c (hunk 3 above)
Comment (ShriramShastry) [reply src/audio/multiband_drc/multiband_drc.c:80]: Ok, Reverted the original code

File: src/audio/multiband_drc/multiband_drc.c [hunk 5]
```diff
@@ -198,17 +199,12 @@ static int multiband_drc_init_coef(struct processing_module *mod, int16_t nch, u
...
 	/* Reset any previous state */
 	multiband_drc_reset_state(&cd->state);
 
 	/* Setup Crossover, Emphasis EQ, Deemphasis EQ, and DRC */
-	ret = multiband_drc_init_coef(mod, channels, rate);
-	if (ret < 0)
-		return ret;
-
-	return 0;
+	return multiband_drc_init_coef(mod, channels, rate);
```
Comment (cujomalainey) [inline src/audio/multiband_drc/multiband_drc.c:207]: This is good, lets keep this

**Review by cujomalainey — APPROVED** [review]

**Review by lyakh — APPROVED** [review]
I don't see where this commit replaces "redundant comments"...

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@wszypelt good to merge ? Looks like CI been pending a while.

**Comment by cujomalainey** [comment]
CI is stuck >5d nows and this is isolated to a module not in default topologies, assuming CI is borked and bypassing

**Pull request merged**